File: jorney/zmq/hello/csharp/server/main/src/ServerMain.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;

using ZeroMQ;

namespace Zmq.Hello
{
	class Server : IDisposable
	{
		private ZContext context;
		private ZSocket channel;

		public static void Main(string[] args)
		{
			using (var server = new Server())
			{
				server.Run();
			}
		}

		public Server()
		{
			context = new ZContext();
			channel = new ZSocket(context, ZSocketType.REP);
		}

		public void Run()
		{
			Initialize();
			Communicate();
		}

		public void Dispose()
		{
			channel.Dispose();
			context.Dispose();
		}

		private void Initialize()
		{
			channel.Bind("tcp://*:5555");
		}

		private void Communicate()
		{
			while (true)
				ProcessRequest();
		}

		private void ProcessRequest()
		{
			var request = Receive();
			Report("Received {0}\n", request);
			Busy();
			Answer("World");
		}

		private string Receive()
		{
			using (var request = channel.ReceiveFrame())
			{
				return request.ReadString();
			}
		}

		private void Answer(string text)
		{
			using (var answer =  new ZFrame(text))
			{
				channel.Send(answer);
			}
		}

		private void Busy()
		{
			Thread.Sleep(500);
		}

		private void Report(string text, params string[] extra)
		{
			Console.Write(text, extra);
		}
	}
}

File: jorney/zmq/hello/cpp/server/main/src/ServerMain.cpp
#include <string>
#include <iostream>
#include "zmq.hpp"
#include "basics/Thread.h"

namespace zmqhello {
	class Server {
	public:
		Server(std::string const& anEndpoint) :
			endpoint(anEndpoint),
			context(1),
			channel(context, ZMQ_REP) {
		}

		void run() {
			initialize();
			communicate();
		}

	private:
		void initialize() {
			std::cout << "Ready to service requests from " << endpoint << " ...\n";
			if (useBroker())
				channel.connect("tcp://" + endpoint);
			else
				channel.bind("tcp://" + endpoint);
		}

		bool useBroker() const {
			return endpoint.find_first_of('*') != 0;
		}

		void communicate() {
			while (true)
				processRequest();
		}

		void processRequest() {
			std::cout << "Received " << receive() << std::endl;
			busy();
			replay("World");
		}

		std::string receive() {
			zmq::message_t request;
			channel.recv(&request);
			return std::string((char*)request.data(), request.size());
		}

		void replay(std::string const& text) {
			zmq::message_t reply(text.size());
			::memcpy(reply.data(), text.c_str(), text.size());
			channel.send(reply);
		}

		void busy() {
			basics::sleep(1);
		}

	private:
		std::string endpoint;
		zmq::context_t context;
		zmq::socket_t channel;
	};
}

int main(int argc, char* argv[]) {
	std::string endpoint = "localhost:5560";
	if (argc > 1)
		endpoint = argv[1];
	zmqhello::Server server(endpoint);
	server.run();
	return 0;
}